iarc_vision/scripts/colortracking.py

```diff
@@ -1,60 +1,298 @@
+from std_msgs.msg import Header
 from cv_bridge import CvBridge, CvBridgeError
 import numpy as np
+import numpy.linalg as la
 import cv2
+import math
+from image_geometry import PinholeCameraModel
+from tf import TransformListener
+from tf.transformations import quaternion_from_euler
 
-class ColorTracker(object):
+
+class ColorTrackerROS(object):
     def __init__(self):
-    	rospy.init_node('color_tracker')
-        self.cv_image = None     # the latest image from the camera
-        self.bridge = CvBridge() # used to convert ROS messages to OpenCV
-        rospy.Subscriber("/usb_cam/image_raw", Image, self.process_image)
-        print "Initializing Color Tracker"
+        rospy.init_node('color_tracker')
+        self.tracker = ColorTracker()
+        self.cv_image = self.processed_image = None  # the latest image from the camera
+        self.boxes = []
+        self.bridge = CvBridge()  # used to convert ROS messages to OpenCV
+        self.cameraModel = PinholeCameraModel()
+        self.tf = TransformListener()
+        rospy.Subscriber("/ardrone/bottom/image_raw", Image, self.process_image)
+        rospy.Subscriber("/ardrone/bottom/camera_info", CameraInfo, self.on_camera_info)
+        print("Initializing Color Tracker")
         cv2.namedWindow('preview_window')
         cv2.namedWindow('binary')
 
+        self.debug_pub = rospy.Publisher("tracker/debug", PoseWithCovarianceStamped, queue_size=10)
+
     def process_image(self, msg):
         """
         Process image messages from ROS and stash them in an attribute
         called cv_image for subsequent processing
-		
+
         """
         try:
-            cv_image = self.bridge.imgmsg_to_cv2(msg, desired_encoding="bgr8")
+            self.cv_image = self.bridge.imgmsg_to_cv2(msg, desired_encoding="bgr8")
             # do image processing here
-            binary_image = cv2.inRange(cv_image, (128,128,128),(255,255,255))
-            self.binary_image = binary_image
-            self.cv_image = cv_image
+            self.boxes, self.processed_image = self.tracker.find_bounding_boxes(self.cv_image, display=False)
+
+            for box in self.boxes:
+                center = np.mean(box, axis=0)
+                heading = get_heading(box, center, self.processed_image)
+                ray = self.cameraModel.projectPixelTo3dRay(center)
+                camera_ray = Vector3Stamped(header=msg.header,
+                                            vector=Vector3(*ray))
+                world_ray = self.tf.transformVector3('map', camera_ray)
+                # print(camera_ray, world_ray)
+                pos, quat = self.tf.lookupTransform('map', msg.header.frame_id, msg.header.stamp)
+                multiplier = -pos[2] / world_ray.vector.z
+                drone_to_roomba = np.array([world_ray.vector.x, world_ray.vector.y, world_ray.vector.z]) * multiplier
+                quat = quaternion_from_euler(0,0,heading)
```

Comment:
Does this also need to account for the orientation of the drone in the world that was discovered above?